### PR TITLE
[stable/datadog] Add fullnameOverride and nameOverride in README and values file

### LINKED ISSUE #11902: [stable/datadog] This just caught us out and did have an impact - it's quite a significant change.
This just caught us out and did have an impact - it's quite a significant change.

The new "fullnameOverride" value isn't mentioned in the default values file or in the table in the README, either, so it's not obvious that it has been introduced and there's no changelog to review prior to applying.

_Originally posted by @nickdgriffin in https://github.com/helm/charts/pull/11250#issuecomment-469301257_

### PR DESCRIPTION
**Pull request by naseemkullah**
#### What this PR does / why we need it:

* Add fullnameOverride and nameOverride in README and values file
* Correct section name `Updating` to `Upgrading` in README 
* Removes blank spaces at EOL in a few places.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #11902

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by naseemkullah** [comment]
cc @hkaj @irabinovitch @xvello @charlyf

**Comment by CharlyF** [comment]
Currently testing, will approve if everything works as expected.

**Review by CharlyF — CHANGES_REQUESTED** [review]
I think a few things are missing in [the initial PR]( https://github.com/helm/charts/pull/11250).
I understand the intent, but it seems strange that using `fullnameOverride` would only override a few things.

Using this custom chart we have:

```
$ helm install --name foo --set fullnameOverride=bar  --set datadog.apiKey=xxx     --set datadog.appKey=xxx    --set clusterAgent.enabled=true     --set clusterAgent.metricsProvider.enabled=true .
NAME:   foo
LAST DEPLOYED: Wed Mar  6 17:09:05 2019
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/ClusterRole
NAME                                       AGE
bar                                        3s
foo-cluster-agent                          3s
foo-cluster-agent-external-metrics-reader  3s

==> v1/ClusterRoleBinding
NAME                                       AGE
bar                                        3s
foo-cluster-agent                          3s
foo-cluster-agent-external-metrics-reader  3s
foo-cluster-agent:system:auth-delegator    3s

==> v1/Deployment
NAME               READY  UP-TO-DATE  AVAILABLE  AGE
foo-cluster-agent  0/1    1           0          3s

==> v1/Pod(related)
NAME                                READY  STATUS             RESTARTS  AGE
bar-4hwfk                           0/1    ContainerCreating  0         3s
foo-cluster-agent-68b758b5fc-q77fr  0/1    ContainerCreating  0         3s

==> v1/RoleBinding
NAME               AGE
foo-cluster-agent  3s


==> v1/Secret
NAME               TYPE    DATA  AGE
bar                Opaque  1     3s
bar-appkey         Opaque  1     3s
foo-cluster-agent  Opaque  1     3s

==> v1/Service
NAME                           TYPE       CLUSTER-IP    EXTERNAL-IP  PORT(S)   AGE
foo-cluster-agent              ClusterIP  10.59.242.56  <none>       5005/TCP  3s
foo-cluster-agent-metrics-api  ClusterIP  10.59.255.80  <none>       443/TCP   3s

==> v1/ServiceAccount
NAME               SECRETS  AGE
bar                1        3s
foo-cluster-agent  1        3s

==> v1beta1/APIService
NAME                             AGE
v1beta1.external.metrics.k8s.io  3s

==> v1beta1/DaemonSet
NAME  DESIRED  CURRENT  READY  UP-TO-DATE  AVAILABLE  NODE SELECTOR  AGE
bar   3        1        0      1           0          <none>         3s
```

Which in my opinion is quite confusing.
Could you align the naming of all objects so they are consistent with the usage of `fullnameOverride` ?

**Comment by naseemkullah** [comment]
Good call @CharlyF , thanks for spotting this.
I made some changes that should address this issue, please have a look when you get a chance.

I was wondering how do you all feel about using recommended labels e.g. https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

I don't mind addressing this in this PR if it's something you would like for the datadog chart.

File: stable/datadog/templates/cluster-agent-deployment.yaml [hunk 1]
```diff
@@ -2,10 +2,10 @@
 apiVersion: apps/v1
 kind: Deployment
 metadata:
-  name: {{ template "datadog.clusterAgent.fullname" . }}
+  name: {{ include "datadog.fullname" . }}-cluster-agent
   labels:
     app: "{{ template "datadog.fullname" . }}"
-    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+    chart: {{ include "datadog.chart" . }}
```
Comment (CharlyF) [inline stable/datadog/templates/cluster-agent-deployment.yaml]: `{{ template "datadog.clusterAgent.fullname" . }}` needs to be updated here: https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-19938b6253581d6defa2db7d73496176R15
and here: 
https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-19938b6253581d6defa2db7d73496176R19
and here:
https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-19938b6253581d6defa2db7d73496176R118

File: stable/datadog/templates/daemonset.yaml [hunk 2]
```diff
@@ -133,7 +133,7 @@ spec:
           - name: DD_CLUSTER_AGENT_AUTH_TOKEN
             valueFrom:
               secretKeyRef:
-                name: {{ template "datadog.clusterAgent.fullname" . }}
+                name: {{ include "datadog.fullname" . }}-cluster-agent
```
Comment (CharlyF) [inline stable/datadog/templates/daemonset.yaml]: This needs to be updated 
https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-c5ed47d16bb168ac09be3d6cdf6b4abaR132

File: stable/datadog/templates/agenthpa-rolebinding.yaml [hunk 3]
```diff
@@ -4,7 +4,7 @@ kind: RoleBinding
 metadata:
   labels:
     app: "{{ template "datadog.fullname" . }}"
-    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+    chart: {{ include "datadog.chart" . }}
```
Comment (CharlyF) [inline stable/datadog/templates/agenthpa-rolebinding.yaml]: `{{ template "datadog.clusterAgent.fullname" . }` has to be changed here https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-b740101d9db451b7817c76af7b448639R10

File: stable/datadog/templates/agent-service-metrics.yaml [hunk 4]
```diff
@@ -2,10 +2,10 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ template "datadog.clusterAgent.fullname" . }}-metrics-api
+  name: {{ include "datadog.fullname" . }}-cluster-agent-metrics-api
   labels:
     app: "{{ template "datadog.fullname" . }}"
-    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+    chart: {{ include "datadog.chart" . }}
```
Comment (CharlyF) [inline stable/datadog/templates/agent-service-metrics.yaml]: The selector with `{{ template "datadog.clusterAgent.fullname" . }}` needs to be updated too https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-c830c155f9433d8d2cb2520cf4b3e680R14

File: stable/datadog/templates/agent-service-cmd.yaml [hunk 5]
```diff
@@ -2,10 +2,10 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ template "datadog.clusterAgent.fullname" . }}
+  name: {{ include "datadog.fullname" . }}-cluster-agent
   labels:
     app: "{{ template "datadog.fullname" . }}"
-    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+    chart: {{ include "datadog.chart" . }}
```
Comment (CharlyF) [inline stable/datadog/templates/agent-service-cmd.yaml]: Needs an update here too https://github.com/helm/charts/pull/11903/commits/e8e3405a850ea2294f73d8557c4329409fed4280#diff-8737d87049f5e1f30969105fb7af06c8R14

**Review by CharlyF — CHANGES_REQUESTED** [review]
I think labels could be a nice addition, but it will for sure be part of another PR. 
I left comments, it appears that the `datadog.clusterAgent.fullname` is still used in a few different spots.
After making the changes (I think I got them all), please try to run the chart with different options (such as `--set clusterAgent.enabled=true` or  `--set clusterAgent.metricsProvider.enabled=true ` to make sure it all works.

**Review by hkaj — COMMENTED** [review]
I'm a little nervous about changing many resource names at once like this. There is little value in it, and it could make the upgrade path not trivial. Please make sure to test not only creating a release, but also upgrading from an older version of the chart. As a reminder, the versioning scheme of this chart doesn't allow breaking changes in minor version bumps.

**Comment by naseemkullah** [comment]
Hi @CharlyF @hkaj  sorry for the huge commit, but this was the only way I saw how to fix the mess I started with the initial PR.

FWIW the chart is now much more standard and using recommended labels 
 which includes `app.kubernetes.io/component` . Short term pain for long term gain IMO.

upgrading with `--force` (to recreate everything) seemed to work for me, so i added this in the Upgrading section of readme.

Overriding fullName also works as expected.

Thoughts?

**Comment by CharlyF** [comment]
Hey @naseemkullah - Thank you very much for your contribution.
As it is unfortunately, I can't accept it as it embeds too many changes on top of a fix.

Please, let's get back to https://github.com/helm/charts/commit/e8e3405a850ea2294f73d8557c4329409fed4280, and include the changes I suggested after in https://github.com/helm/charts/pull/11903#pullrequestreview-211797663.

As per my previous comment:
> I think labels could be a nice addition, but it will for sure be part of another PR.
For now, let make sure all users of our chart have a seamless experience upgrading, in a later PR we can review adding standards and use recommended labels. 

For now, let's scope this change to fix the initial issue. 

Thanks again!
.C

**Comment by naseemkullah** [comment]
Oops, sorry. Added it back @CharlyF !

**Pull request merged**